src/transformStateWithClones.js

```
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = Object.keys(currentState).reduce((newState, key) => {
+          if (!action.keysToRemove.includes(key)) {
+            newState[key] = currentState[key];
+          }
+
+          return newState;
+        }, {});
+        break;
+      default:
+        break;
```

Default case does not perform observable error handling. The checklist requires: "The switch statement must include a default case that handles unexpected action.type values for error handling" and "In the switch default case, perform error handling (e.g., throw an error or skip) ... The behavior must be observable in code." (2.4 and 3.3). Currently the default simply `break`s which silently ignores invalid action types. Consider throwing an Error with a clear message, e.g.: `throw new Error('Unknown action type: ' + action.type);` to make failures visible to callers/tests.